### ISSUE TITLE
Password caching is terrifying

### ISSUE THREAD

**Issue by djspiewak** [sbt/sbt-pgp#68]
As per discussion…

I think this is a problem that we need to solve at the SBT level, since it's actually SBT that controls the `SecurityManager`.  Fixing this problem in a general way at the SBT level would also give us a general tool for use in other areas, such as `credentials` and stuff like my [1Password trick](https://gist.github.com/djspiewak/d8082be3efeb875acd4d).

A solution at the SBT level should look something like this:

``` scala
trait SecureCache {
  def put[Client: ClassTag](key: Key)(value: key.Value): Unit
  def get(key: Key): Option[key.Value]

  trait Key {
    type Value

    final override def hashCode = System.identityHashCode(this)
    final override def equals(that: Any) = this eq that
  }
}
```

I'm using dependent method types here just to make the API slightly nicer on the client side, since tossing an extra type parameter in there would require over-specification of type information.  Lots of stuff you can do there (also, if you want to be gross, you can always take a second `ClassTag` and use a `String` key).

Anyway, the `Client` tag is taken by the `put` function to restrict access.  Whenever the `get` method is called, we should interrogate `Thread.currentThread.getStackTrace` in order to ensure that we're actually accessing the data from either the `Client` class or an inner class thereof (including lambdas).  Note that we need to look at the _immediately_ previous stack frame entry and validate the access there (hence the constraint about inner classes).  If we just walk arbitrarily far up the stack trace, looking for _any_ instance of `Client`, we open up a security hole.

Here's where the `SecurityManager` comes in…  Reflective access _of any kind_ to the implementation of `SecureCache` must be prevented.  Furthermore, the SBT `SecurityManager` needs to forbid installation of custom security managers which could override these restrictions (it doesn't currently, which actually allows plugins to do devious things with the current security manager).  Note that `SecureCache` can and probably should do some extra checking inside of `put` to ensure that it is _specifically_ SBT's `SecurityManager` that is installed, bailing out without storing the data if the check is failed.  We don't want someone doing something tricky by forking a subprocess instance of SBT with a custom security manager that overrides SBT's default.

Obviously this isn't fool-proof.  There are plenty of threat vectors which this doesn't prevent.  Heck, just taking a heap dump (can you forbid that with a `SecurityManager`?) seems like a pretty direct way to read this information, but trickier things like core dumps also seem like a possibility.  At the end of the day, SBT is a build tool, and build tools are all about forking arbitrary and unrestricted subprocesses.  We can make things _harder_ though, and that's all security is about.

So to summarize, `SecureCache` provided by SBT as part of the core framework and protected by SBT's `SecurityManager` would give plugins (and builds!) the ability to stash data in memory in such a way that it is very hard (but not impossible) to exfiltrate.  We could probably also offer some nice APIs on top of this for doing things like prompting for passwords and automatically storing them in `SecureCache`, but that's just gravy.  Having `SecureCache` in the first place is the important bit.

**Comment by jsuereth** [comment]
@djspiewak Do you have a full implementation of this?   We're starting to pull out sbt modules for a 1.0 and this would be a good thing to design/implement/get in there.  cc> @eed3si9n

**Comment by djspiewak** [comment]
I do not even have a partial implementation. :-( The trickiest thing I think is getting the security manager to do the right thing.  I'm not 100% certain that it's going to allow us to block off access in precisely the manner I describe.

I'm a bit busy for the next few weeks, but I can poke around with it afterward.  The rest of the design is fairly trivial; it's really just the security that's hard/unknown.

**Comment by wsargent** [comment]
So from http://stackoverflow.com/questions/27232591/how-to-use-public-key-openpgpgpg-without-access-to-physical-key-ring-in-java it looks like Bouncycastle requires direct access to the keyring, so that's a problem.

However!  It's possible to call out to gpg directly, and leverage gpg-agent without going through bouncycastle, using gnupg-for-java.

https://github.com/guardianproject/gnupg-for-java

It's a JNI integration, but for the use case, it looks like the best option -- combined with a Yubikey 4 or Neo, you can set up GPG keys appropriately, and use Yubikey's access to pinentry to keep the root PGP keys completely offline.  Best guide for Yubikey / GPG set up is here:

https://github.com/drduh/YubiKey-Guide

**Comment by eed3si9n** [comment]
useGpg is now default, so I am closing this.
